### PR TITLE
Feat/vexflow4

### PR DESCRIPTION
**Pull request by rvilarl**
develop merged
vexflow updated to 4.1.0-alpha.0

### CODE REVIEW

**Comment by sschmidTU** [comment]
@rvilarl Looks good, just one issue:

I'm getting an error `BadGlyph: Glyph noteheadCircleX does not exist in font.
    at Glyph.lookupGlyph (glyph.js:120:1)`
with OSMD Function Test - Notehead Shapes, and the notes don't render.
They should look like this:
image

Same in `test_percussion_display_step_from_xml.musicxml`.

Also, as with Vexflow 1.2.93 before we patched it out, the Chrome console gives a warning `Canvas2D: Multiple readback operations using getImageData are faster with the willReadFrequently attribute set to true. See: https://html.spec.whatwg.org/multipage/canvas.html#concept-canvas-will-read-frequently`.

We patched that in Vexflow 1.2.93 in https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/issues/1242.
Basically, it was this line in renderer.js:
`this.ctx = Renderer.bolsterCanvasContext(this.element.getContext('2d', {willReadFrequently: true}));`

You could either set that as a default as well, or offer an option. There seems to be no downside to this, though the performance boost also seems to be minimal.

I can also merge before these issues are resolved, but then I'll add this to the issue list in #1139.

**Comment by rvilarl** [comment]
@sschmidTU vexflow updated and develop merged.

**Comment by sschmidTU** [comment]
@rvilarl The CI build (Appveyor) gets a lot of errors, any idea why?

like these:
```console
ERROR in C:\projects\opensheetmusicdisplay\src\MusicalScore\Graphical\VexFlow\VexFlowMeasure.ts
./src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts 420:47-55
[tsl] ERROR in C:\projects\opensheetmusicdisplay\src\MusicalScore\Graphical\VexFlow\VexFlowMeasure.ts(420,48)
      TS2304: Cannot find name 'position'.
 @ ./src/MusicalScore/Graphical/VexFlow/VexFlowGraphicalSymbolFactory.ts 3:0-50 40:19-33 66:42-56
 @ ./src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts 2:0-80 31:49-78
 @ ./src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts 2:0-110 149:25-52 674:103-130 868:99-126
 @ ./demo/index.js 1:0-91 437:36-57
ERROR in C:\projects\opensheetmusicdisplay\src\MusicalScore\Graphical\VexFlow\VexFlowMusicSheetCalculator.ts
./src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts 1187:37-40
[tsl] ERROR in C:\projects\opensheetmusicdisplay\src\MusicalScore\Graphical\VexFlow\VexFlowMusicSheetCalculator.ts(1187,38)
      TS2[304](https://ci.appveyor.com/project/sebastianhaas/opensheetmusicdisplay/builds/45770915#L304): Cannot find name 'Vex'.
 @ ./src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts 2:0-110 149:25-52 674:103-130 868:99-126
 @ ./demo/index.js 1:0-91 437:36-57
```

**Comment by rvilarl** [comment]
Perhaps I forgot to commit files. I will look into it.

**Comment by rvilarl** [comment]
Yes I forgot some. :( Let us see now, I cannot wait for the build to end. Happy Xmas!

**Comment by rvilarl** [comment]
@sschmidTU any reason not to merge it?

**Comment by sschmidTU** [comment]
@rvilarl No, I just lost sight of it during the holidays, will take a look these days :) thanks for the reminder.

**Comment by sschmidTU** [comment]
@rvilarl This is just merging osmd's develop and updating vexflow in the package.json, right? Any code changes to adjust to the new vexflow version?
Ideally, I would like to merge develop first and then update the vexflow version, at least then it would be easier to see the changes specifically caused by the vexflow update.

But no problem, I will simply generate blessed images without the vexflow update, then create visual diffs for the vexflow update. Then I'll merge regardless, but if there are any new issues, I'll note them in #1139.

**Comment by sschmidTU** [comment]
New/Changed issues in Vexflow 4.2.0-beta.0 (PR #1247):

- [ ] The last notes of a measure are now even further from the measure end, the margin is way too large in my view:
4.0.3:
image

4.2.0-beta.0:
image

diff:
image

- [ ] Scores are again much larger than in 4.0.3 (which was already larger than 1.2.93), probably because measures need more space. In our default piece (Beethoven Geliebte), 4 measures used to fit into row/system 2, now only 3 fit.
A good example of a much larger score is `JohannSebastianBach_PraeludiumInCDur` (3116 -> 4201 pixels high)

This is very obvious in `test_tuplet_consecutive_3and5_tuplets`. Is there maybe an issue with triplets using a lot of space?
image

image

or this one:
image

Strangely, almost no size change in `test_tuplet_placement`
image

There is definitely a certain mechanic at work here that changes the measure size a lot. Not sure how much on OSMD's or Vexflow's side.

- [ ] tab sample `BrookeWestSample` errors and doesn't render:
```console
Error: [RuntimeError] TickMismatch: Voices should have same total note duration in ticks.
StackTrace: 
Error: [RuntimeError] TickMismatch: Voices should have same total note duration in ticks.
    at formatter.js:255:1
    at Array.reduce (<anonymous>)
    at Formatter.getResolutionMultiplier (formatter.js:253:1)
    at createContexts (formatter.js:24:1)
    at Formatter.createTickContexts (formatter.js:296:1)
    at Formatter.preCalculateMinTotalWidth (formatter.js:213:1)
    at VexFlowMusicSheetCalculator.calculateMeasureXLayout (VexFlowMusicSheetCalculator.ts:205:40)
```

Also a simpler tab sample: `OSMD_Function_Test_Tablature_Vibrato`, `_Alleffects`, `_Bends`, etc

I think we VexFlowPatched this to just ignore these mismatches, maybe specifically in tabs, it caused zero issues to just ignore the mismatch. I can look into this and why it doesn't work anymore with the current PR changes.

- [x] Improvement: drum clef now aligned with other clefs
image

image

image

- [x] Improvement: grace note better positioned
image

- [x] concern, but not critical yet: note distances sometimes very close, almost no padding
image

image
(see Upmordent)

- [ ] Voices not aligned anymore in `OSMD_Function_Test_Voice_Alignment`
image

image

image

- [ ] Rehearsal mark box around letter doesn't render in `test_rehearsal_marks_bolivia`. We had VexflowPatched this. Maybe I can look into this.
image

image

image

more simply in `test_rehearsal_marks_simple_one_measure`:
image

image

- [x] Improvement: `test_string_number_collisions`
image

image

- [ ] Consecutive Tuplet number hiding doesn't work anymore:
image

image

see the variables `TupletNumberLimitConsecutiveRepetitions`, `TupletNumberMaxConsecutiveRepetitions`, etc. in EngravingRules

- [ ] Pedal brackets rendering causes errors, so we have to disable this in this PR via `osmd.EngravingRules.RenderPedals = false` to for example be able to render our default piece, Beethoven Geliebte. I can look into this, we have a VexFlowPatch for pedalmarking which adds a function setEndStave(). Would be nice to get this into vexflow via a PR, I'll see if I can do that.

```console
Error loading sheet: TypeError: pedalMarking.setEndStave is not a function
StackTrace: 
TypeError: pedalMarking.setEndStave is not a function
    at VexFlowPedal.getPedalMarking (VexFlowPedal.ts:126:35)
```

This is a quick look, I think we can take it from here and work on these first. It's quite a bit to fix, but we can take it one at a time, step by step!

I'll still merge this so that we have a clean slate to work from (and clean diffs moving forward).
